db/flush_job.cc

```diff
@@ -306,6 +317,272 @@ void FlushJob::Cancel() {
+      }
+    }
+
+    // Check status and propagate
+    // potential error status from c_iter
+    if (!s.ok()) {
+      c_iter.status().PermitUncheckedError();
+    } else if (!c_iter.status().ok()) {
+      s = c_iter.status();
+    }
+
+    // Range tombstone transfer.
+    if (s.ok()) {
+      auto range_del_it = range_del_agg->NewIterator();
+      for (range_del_it->SeekToFirst(); range_del_it->Valid();
+           range_del_it->Next()) {
+        auto tombstone = range_del_it->Tombstone();
+        new_first_seqno =
+            tombstone.seq_ < new_first_seqno ? tombstone.seq_ : new_first_seqno;
+        s = new_mem->Add(
+            tombstone.seq_,        // Sequence number
+            kTypeRangeDeletion,    // KV type
+            tombstone.start_key_,  // Key is start key.
+            tombstone.end_key_,    // Value is end key.
+            nullptr,               // KV protection info set as nullptr since it
+                                   // should only be useful for the first add to
+                                   // the original memtable.
+            false,                 // : allow concurrent_memtable_writes_
+                                   // Not seen as necessary for now.
+            nullptr,               // get_post_process_info(m) must be nullptr
+                      // when concurrent_memtable_writes is switched off.
+            nullptr);  // hint, only used when concurrent_memtable_writes_
+                       // is switched on.
+
+        if (!s.ok()) {
+          break;
+        }
+
+        // If new_mem is full, add it to purged_mems, and allocate new
+        // memtable.
+        if (new_mem->ShouldScheduleFlush()) {
+          // Rectify the first sequence number, which (unlike the earliest seq
+          // number) needs to be present in the new memtable.
+          new_mem->SetFirstSequenceNumber(new_first_seqno);
+          purged_mems.push_back(new_mem);
+          new_mem = new MemTable(
+              (cfd_->internal_comparator()), *(cfd_->ioptions()),
+              mutable_cf_options_, cfd_->write_buffer_mgr(),
+              mems_[0]->GetEarliestSequenceNumber(), cfd_->GetID());
+          new_first_seqno = kMaxSequenceNumber;
+        }
+      }
+    }
+
+    // If everything happened smoothly and new_mem contains valid data,
+    // decide if it is flushed to storage or kept in the imm()
+    // memtable list (memory).
+    if (s.ok() && (new_first_seqno != kMaxSequenceNumber)) {
+      // Rectify the first sequence number, which (unlike the earliest seq
+      // number) needs to be present in the new memtable.
+      new_mem->SetFirstSequenceNumber(new_first_seqno);
+
+      // The new_mem is added to the list of immutable memtables
+      // only if it filled at less than 10% capacity (arbitrary heuristic).
+      if (new_mem->ApproximateMemoryUsage() <
+          static_cast<size_t>(
+              ceil(0.1 * mutable_cf_options_.write_buffer_size))) {
```

Comment:
Arbitrary metric: add to imm() only if memtable at less then 10% capacity. Reason for this: minimize Get/Read overheads that come from storing an extra Imm memtable, while giving us a change to perform in memory compaction.

Reply:
Btw, could avoid floating point with

    (mutable_cf_options_.write_buffer_size + 9) / 10

Reply:
Nice trick, I'll go ahead and edit that (even though long term there is a chance we need the flexibility a double would bring).